### ISSUE TITLE
Triggering Agents by events

### ISSUE THREAD

**Issue by Azimoth** [huginn/huginn#2221]
Hi, I try to use my current location in Huginn. So far my setup works... in theory.

Huginn receives the coordinates using the MQTT Agent and then hands them over to a Commander Agent that configures the coordinates for the DarkSky Agent.

The problem is triggering the agents.
So far I tried running the MQTT Agent once a minute, but I want to run the Commander Agent only, if new coordinates are received. If the Commander Agent runs without having received new coordinates, it sets location to {"location"=>","}, instead of keeping the old coordinates. This in turn leads (of course) to an error from the DarkSky agent.

Is there a way to trigger the Commander Agent from the MQTT Agent "on change"?

**Comment by dsander** [comment]
I am not too familiar with the MQTT Agent, it emits an empty Event if the location did not change?

The `ChangeDetectorAgent` to filter out duplicates or the `TriggerAgent` to only forward the Event if certain conditions are met might help.

**Comment by Azimoth** [comment]
No, the LocationAgent emits an empty location. The only way I found to trigger the LocationAgent was by schedule, so if it runs every ten minutes, but there was no location received by MQTT in the mean time the LocationAgent emits ",", even though there are still coordinates in the MQTT agents memory.

**Comment by dsander** [comment]
When you run the `MqttAgent` on a schedule it checks the configured topic for messages and creates one event per message (I assume the broker removes messages once they have been delivered?). It checks if the message is identical to the previous message if the `retain` or `duplicate` attributes are set on the "packet", but I don't know how those can be set.

The simplest solution to avoid the empty location events would probably be a `TriggerAgent` that forwards the Event to your commander only if the message does not equal `,`.

**Comment by Azimoth** [comment]
This sounds about right for me:

- To create a CommanderAgent that receives an event from a WeatherAgent every morning to kick an agent flow that is only useful in a nice weather, try this:
{% if conditions contains 'Sunny' or conditions contains 'Cloudy' %} run{% endif %}

But apparently 

"{% if conditions contains 'Sunny' or conditions contains 'Cloudy' %} run{% endif %}"

is not the right syntax, as simply copying it over leads to the message "Sorry, there appears to be an error in your JSON input.  Please fix it before continuing."
I basically want to say "if locations != "," trigger LocationsAgent". 
Where would I find a working example for this, that I could adjust to my needs?

**Comment by dsander** [comment]
This JSON works for me:

```json
{
  "action": "{% if conditions contains 'Sunny' or conditions contains 'Cloudy' %} run{% endif %}"
}
```

**Comment by Azimoth** [comment]
It does indeed. I simply did not know I had to put it next to `"action":`. Thx!^^

**Comment by Azimoth** [comment]
I just noticed that I was causing the problem that made me start this thread myself: I thought I had to trigger the LocationAgent expressively and by running it more often than new coordinates were received I made it overwrite older coordinates.
During my try to set up a CommandAgent to control this, I noticed, that it is triggered automatically when the MQTT Agent received the coords... Had I left it alone in the first place, everything would've been fine. -.-'

**Comment by dsander** [comment]
Your "LocationAgent" is a MqttAgent running on a schedule, right? That should update the "Last Check" column.

**Comment by Azimoth** [comment]
No, the MQTT Agent only receives the coordinates. The LocationAgent is a Commander Agent (see first post) that is configured like this:
```
{
    "action": "configure",
    "configure_options": {
      "location": "{{message.lon}},{{message.lat}}"
    }
}
```

**Comment by dsander** [comment]
If the `CommanderAgent` does not run on a schedule it will not update "Last Check". after it received a Event it should update "Last received at".
